### PR TITLE
chore(sinks): Only log `200`-`299` responses as successful

### LINKED ISSUE #2426: 404 responses should not be considered successful
We have a user in chat that shared the following log line:

```
Apr 23 15:29:41.321 TRACE sink{name=datadog_metrics_internal_metrics type=datadog_metrics}:request{request_id=1}: vector::sinks::util::sink: Response successful. response=Response { status: 404, version: HTTP/1.1, headers: {"date": "Thu, 23 Apr 2020 15:29:41 GMT", "content-type": "text/html; charset=utf-8", "content-length": "0", "connection": "keep-alive", "pragma": "no-cache", "cache-control": "no-cache", "x-dd-version": "35.2425337", "dd-pool": "dogweb", "content-security-policy": "frame-ancestors 'self'; report-uri https://api.datadoghq.com/csp-report", "x-frame-options": "SAMEORIGIN", "x-content-type-options": "nosniff", "strict-transport-security": "max-age=15724800;"}, body: b"" }
```

Notice that the response is "successful" yet we received a 404. That doesn't make sense.

### PR DESCRIPTION
**Pull request by ktff**
Closes #2426

Fix for all `http` sinks that were using `ServiceSink`.

### CODE REVIEW

File: src/sinks/util/service.rs [hunk 1]
```diff
@@ -163,7 +164,7 @@ pub struct TowerRequestLayer<L, Request> {
 impl<S, L, Request> tower::layer::Layer<S> for TowerRequestLayer<L, Request>
 where
     S: Service<Request> + Send + Clone + 'static,
-    S::Response: Send + 'static,
+    S::Response: Send + 'static + Response,
```
Comment (bruceg) [inline src/sinks/util/service.rs:167]: The lifetime should go at the end, and probably this trait should go at the beginning before things like `Send`.

File: src/sinks/util/sink.rs [hunk 2]
```diff
@@ -706,6 +709,36 @@ where
...
+    fn is_successful(&self) -> bool {
+        true
+    }
+}
+
+impl Response for () {}
+impl<'a> Response for &'a str {}
+
+#[cfg(feature = "rusoto_kinesis")]
+impl Response for rusoto_kinesis::PutRecordsOutput {}
```
Comment (bruceg) [inline src/sinks/util/sink.rs:724]: Move this into the `aws_kinesis_streams` sink to avoid the `#[cfg(X)]` here,

File: src/sinks/util/sink.rs [hunk 3]
```diff
@@ -706,6 +709,36 @@ where
...
+#[cfg(feature = "rusoto_s3")]
+impl Response for rusoto_s3::PutObjectOutput {}
```
Comment (bruceg) [inline src/sinks/util/sink.rs:726]: …and this into the `aws_s3` sink,

File: src/sinks/util/sink.rs [hunk 4]
```diff
@@ -706,6 +709,36 @@ where
...
+#[cfg(feature = "rusoto_firehose")]
+impl Response for rusoto_firehose::PutRecordBatchOutput {}
```
Comment (bruceg) [inline src/sinks/util/sink.rs:728]: …and this into the `aws_kinesis_firehouse` sink.

File: src/sinks/util/sink.rs [hunk 5]
```diff
@@ -706,6 +709,36 @@ where
...
+    fn is_successful(&self) -> bool {
+        self.status().is_success()
+    }
+}
+
+impl<T: fmt::Debug> Response for http02::response::Response<T> {
+    fn is_successful(&self) -> bool {
+        self.status().is_success()
+    }
+}
```
Comment (bruceg) [inline src/sinks/util/sink.rs:740]: These could go into `src/sinks/util/http.rs` and `src/sinks/util/http2.rs` respectively, though I don't think it's a big deal.

**Comment by bruceg** [comment]
What do people think about merging this approach with `RetryLogic` such that `should_retry_response` could just call a method in this new `trait Response`? We probably wouldn't want to generate a string where it isn't needed, which could be resolved in a few different ways:

1. Have `enum RetryAction` encapsulate a `FnOnce` that generates and returns the string when it is actually. I have a suspicion though that this wouldn't actually be more efficient than just generating the string and throwing it away if not needed.
2. Have two different methods, one returns just the state (possibly just the bool like here) and the other returns the enum with the strings. This adds extra code.
3. Have two methods, one returns the state with no strings, and one returns *only* the message if it's needed. More code again.

**Comment by ktff** [comment]
Moving `should_retry_response` to `Response` would make it hard to have different/upgraded retry logic for the same response. We have one such case, `ClickhouseRetryLogic` reuses `HttpRetryLogic` and adds some of it's own logic for the same response type.

Regarding the `String`, I think the potential performance isn't worth the mentioned cost.

**Comment by bruceg** [comment]
> Moving `should_retry_response` to `Response` would make it hard to have different/upgraded retry logic for the same response type. We have one such case, `ClickhouseRetryLogic` reuses `HttpRetryLogic` and adds some of it's own logic for the same response type.

Yeah, I had thought about that and had noticed the same case. It would be necessary to introduce a newtype wrapper for clickhouse. This would allow for the separate impl, but requires a bit of extra code.

> Regarding the `String`, I think the potential performance isn't worth the mentioned cost.

Thinking about it, you're probably right. The strings would only be created when a response indicates a failure, which is already not the fast path.

**Pull request merged**